nest/src/token/pUSD.sol

```diff
@@ -6,137 +6,493 @@ import { Initializable } from "@openzeppelin/contracts-upgradeable/proxy/utils/I
+     * @return shares Amount of shares minted
+     */
+    function deposit(
+        uint256 assets,
+        address receiver,
+        address controller,
+        uint256 minimumMint
+    ) public virtual nonReentrant returns (uint256 shares) {
+        if (receiver == address(0)) {
+            revert InvalidReceiver();
+        }
+
+        ITeller teller = ITeller(address(_getpUSDStorage().boringVault.teller));
+
+        // Verify deposit is allowed through teller
+        if (teller.isPaused()) {
+            revert TellerPaused();
+        }
+        if (!teller.isSupported(IERC20(asset()))) {
+            revert AssetNotSupported();
+        }
+
+        // Transfer assets from sender to this contract first
+        SafeERC20.safeTransferFrom(IERC20(asset()), msg.sender, address(this), assets);
+
+        // Then approve teller to spend assets using forceApprove
+        SafeERC20.forceApprove(IERC20(asset()), address(teller), assets);
+
+        // Deposit through teller
+        shares = teller.deposit(
+            IERC20(asset()), // depositAsset
+            assets, // depositAmount
+            minimumMint // minimumMint
+        );
+
+        // Transfer shares to receiver
+        _mint(receiver, shares);
+
+        emit Deposit(msg.sender, receiver, assets, shares);
+        return shares;
+    }
+
+    /**
+     * @notice Burn shares and withdraw corresponding assets
+     * @param shares Amount of shares to burn
+     * @param receiver Address that will receive the assets
+     * @param controller Address that currently controls the shares
+     * @return assets Amount of assets withdrawn
+     */
+    function redeem(
+        uint256 shares,
+        address receiver,
+        address controller,
+        uint256 price,
+        uint64 deadline
+    ) public virtual nonReentrant returns (uint256 assets) {
+        if (receiver == address(0)) {
+            revert InvalidReceiver();
+        }
+        if (controller == address(0)) {
+            revert InvalidController();
+        }
+        if (deadline < block.timestamp) {
+            revert("Deadline expired");
+        }
+
+        // Get AtomicQueue from storage
+        IAtomicQueue queue = _getpUSDStorage().boringVault.atomicQueue;
+
+        // Create AtomicRequest struct
+        IAtomicQueue.AtomicRequest memory request = IAtomicQueue.AtomicRequest({
+            deadline: deadline,
+            atomicPrice: uint88(price),
+            offerAmount: uint96(shares),
+            inSolve: false
+        });
+
+        IERC20(address(this)).safeIncreaseAllowance(address(queue), shares);
+
+        // Update atomic request
+        queue.updateAtomicRequest(ERC20(address(this)), ERC20(asset()), request);
+
+        // Get assets received from vault
+        assets = shares; // 1:1 ratio for preview to match actual redemption
+        IERC20(asset()).safeTransfer(receiver, assets);
```

Comment:
Should we wait until the request is fully processed before giving the assets back to the user? Or are we confident as soon as the request is submitted that the assets will become available?